File: P1.py
# TC : O(n log n)
# SC : O(1)
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums1.sort()
        nums2.sort()
        if len(nums1) > len(nums2):
            return self.intersect(nums2, nums1)
        l = 0
        h = len(nums2) - 1
        res = []
        left_ind = 0
        for i in nums1:
            ind = self.binary_search(i, h, left_ind, nums2)
            if ind!=-1:
                h -= 1
                nums2.remove(i)
                res.append(i)
        return res
    def binary_search(self, i, h, left_ind, nums1):
        index = -1
        while left_ind <= h:
            mid = left_ind + (h - left_ind)//2
            if nums1[mid] == i:
                index = mid
                h = mid - 1
            elif nums1[mid] > i:
                h = mid - 1
            else:
                left_ind = mid + 1
        return index
        
        
        

File: p1.java
// Time Complexity :O(m+n)
// Space Complexity : O(min(m,n))
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach



class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        //Check which array is short and proceed accordingly
        if(nums1.length > nums2.length){
            return intersect(nums2, nums1);
        }

        List<Integer> result = new ArrayList<>();

        //Store numbers of short array into hashmap with it's frequency
        HashMap<Integer, Integer> map = new HashMap<>();

        for(int num: nums1){
            map.put(num,map.getOrDefault(num,0)+1);
        }

        //Go through all the numbers in 2nd array to find common numbers
        for(int num: nums2){
            if(map.containsKey(num)){
                result.add(num);
                int temp = map.get(num);
                if(temp == 1){
                    map.remove(num);
                }
                else{
                    map.put(num, temp-1);
                }
            }
        }

        //Put all the common numbers in array
        int size= result.size();
        int[] res = new int[size];
        for(int i=0; i<size; i++){
            res[i] = result.get(i);
        }

        return res;
    }
}